ipn/ipnlocal/local.go

```diff
@@ -6023,3 +6025,138 @@ func mayDeref[T any](p *T) (v T) {
+			ipp, err := netip.ParseAddrPort(candidate.DERP())
+			if err != nil {
+				continue
+			}
+			if ipp.Addr() == tailcfg.DerpMagicIPAddr {
+				regionID := int(ipp.Port())
+				if !slices.Contains(sortedRegions, regionID) {
+					sortedRegions = append(sortedRegions, regionID)
+				}
+				candidateLatencyMap[regionID] = append(candidateLatencyMap[regionID], candidate)
+			}
+		} else {
+			if candidate.Hostinfo().Location() != nil {
+				candidateCoordinates := []float64{candidate.Hostinfo().Location().Latitude, candidate.Hostinfo().Location().Longitude}
+				distance := longLatDistance(derpHomeCoordinates, candidateCoordinates)
+				candidateDistanceMap[distance] = append(candidateDistanceMap[distance], candidate)
+				distances = append(distances, distance)
+			}
+		}
+	}
+	sortedRegions = sortRegions(sortedRegions, lastReport)
+	if len(sortedRegions) > 0 && candidateLatencyMap[sortedRegions[0]] != nil {
+		preferredExitNodeID = candidateLatencyMap[sortedRegions[0]][0].StableID()
+		preferredExitNodeName = candidateLatencyMap[sortedRegions[0]][0].Name()
+	} else {
+		slices.Sort(distances)
+		if len(distances) > 0 {
+			if len(candidateDistanceMap[distances[0]]) == 1 {
+				preferredExitNodeID = candidateDistanceMap[distances[0]][0].StableID()
+				preferredExitNodeName = candidateDistanceMap[distances[0]][0].Name()
+			} else {
+				chosen := pickWeighted(candidateDistanceMap[distances[0]])
+				if chosen.Valid() {
+					preferredExitNodeID = chosen.StableID()
+					preferredExitNodeName = chosen.Name()
+				}
+			}
+		}
+	}
+	if preferredExitNodeID.IsZero() {
+		return preferredExitNodeID, "", errNoExitNodes
+	}
+	return preferredExitNodeID, preferredExitNodeName, nil
+}
+
+func pickWeighted(candidates []tailcfg.NodeView) tailcfg.NodeView {
+	maxWeight := 0
+	var chosenCandidate tailcfg.NodeView
+	for _, candidate := range candidates {
+		if candidate.Hostinfo().Location() != nil && candidate.Hostinfo().Location().Priority > maxWeight {
+			maxWeight = candidate.Hostinfo().Location().Priority
+			chosenCandidate = candidate
+		}
+	}
+	return chosenCandidate
+}
+
+// sortRegions returns a list of sorted regions by ascending latency given a list of region IDs and a netcheck report.
+func sortRegions(regions []int, lastReport *netcheck.Report) []int {
+	sort.Slice(regions, func(i, j int) bool {
+		iLatency, ok := lastReport.RegionLatency[regions[i]]
+		if !ok || iLatency == 0 {
+			iLatency = math.MaxInt
+		}
+		jLatency, ok := lastReport.RegionLatency[regions[j]]
+		if !ok || jLatency == 0 {
+			jLatency = math.MaxInt
+		}
+		return iLatency < jLatency
+	})
+	return regions
+}
+
+// longLatDistance returns an estimated distance given two lists of float64 values that represent polar coordinates.
+// The first index is latitude and the second index is longitude.
+// Value is returned in meters. We tolerate up to 100km difference in accuracy.
+func longLatDistance(firstDistance []float64, secondDistance []float64) float64 {
```

Comment:
If we only want to know which distances are bigger/smaller can we do away with all the math.Pi / 180 s and much of the trig stuff? could we do just `sqrt(diffLat^2 + diffLong^2)` ? (or even just diffLat+diffLong? how does math work?)

Reply:
And are they polar coords or latitude, longitude pairs? because those are different right?

And if they're latitude, longitude pairs does this algo work around wherever the zero lines are?

Reply:
used haversine formula. geographic coordinates.

Reply:
ok, yep, looking at the call site I see we're calling with latitude and longitude, so the variables shouldn't be called 'distance' I think? seems confusing?

Reply:
Does this formula work for eg 179 degrees longitude to -179 degrees longitude? Maybe we should add some tests?

Reply:
Latitude and longitude needs the fancy math because one degree of longitude is a different linear distance depending on latitude. I believe one degree of latitude is always the same linear distance.

This is already the simplified version that assumes the world is a sphere (it is not).